### PR TITLE
Ext_proc: restrict mode_override to header response only

### PR DESCRIPTION
**Pull request by yanjunxiang-google**
This PR is to address: https://github.com/envoyproxy/envoy/issues/29302

Currently, ext_proc filter supports mode_override when ext_proc server sending response back to Envoy:
https://github.com/envoyproxy/envoy/blob/0e8074b0a9cff3fa1ed0728aa6375433f25d88c1/api/envoy/service/ext_proc/v3/external_processor.proto#L174

While this does provide capability to ext_proc, like skip the rest of the data processing when header already contains sufficient info. However, this does make the state machine very complicated.  The big question to answer is what if the data mode is STREAMED, and there are five gRPCs for body already sent out.  Then the body response for the 1st body chunk want to change the mode to NONE. Then the 2nd response want to change it to BUFFERED_PARTIAL, then the 3rd response want to change it back to STREAMED, ..., and so on.

There are a lot of design questions we need to answer if we allow the state machine can transit from one state to any random state when there are outstanding gRPC calls in the queue, and more data is coming in to that queue.

To simplify this, the solution is to restrict the mode override to be only applicable to the state when Envoy is waiting for header response.  Since everything is paused at such state, allow mode_override in this state is relatively safe and easy to support.

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/filters/http/ext_proc/processor_state.h
Comment (yanjunxiang-google) [inline source/extensions/filters/http/ext_proc/processor_state.h:31]: No need to maintain this delivered flag if we do not support mode changing after streaming data to ext_proc server already started.

So basically,  there are two data modes which may put data into the chunk_queue:
1) in STREAMED mode, every chunk in the queue is delivered.
2) in BUFFERED_PARTIAL mode, the first chunk in the queue is delivered. And the rest of the chunks is not delivered, and no need to be delivered.

BTW, delivered means the data is sent to the ext_proc server.

File: source/extensions/filters/http/ext_proc/processor_state.h
Comment (yanjunxiang-google) [inline source/extensions/filters/http/ext_proc/processor_state.h:78]: This callback state can be removed since we are going to restrict mode_override to header response only. All the code corresponding to this callback state will be  cleaned up.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (yanjunxiang-google) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:332]: We are splitting sendBodyChunk() into two operations: 
1) setupBodyChunk(), which is copy the data into message  2) sendBodyChunk()

The reason is that unit test fuzzer detected a problem with sendBodyChunk, it will call stream_->send(), which will automatically call onReceiveMessage(), then handleBodyResonse(), then dequeueStreamingChunk(). However at that moment, the enqueueStreamingChunk() is not called yet, so, nothing in the queue, and the body response is not handled.  
Thus, the change is to call the enqueueStreamingChunk() before sendBodyChunk().  As enqueueStreamingChunk() also moves the data to the queue buffer. i.e, the data will become empty after enqueueStreamingChunk() here. So, need to call setupBodyChunk() before it to setup the request properly.

**Comment by yanjunxiang-google** [comment]
Kind Ping!

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (yanjunxiang-google) [inline source/extensions/filters/http/ext_proc/processor_state.cc:260]: This while loop is removed since there will be no mode change during data processing. So, for STREAMED, mode, the gRPC response is always for the first chunk in the queue.

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (yanjunxiang-google) [inline source/extensions/filters/http/ext_proc/processor_state.cc:325]: As the leftover data buffer is stored in received_data_, we can inject it back to filter chain in one shot. No need this while loop any more.

File: source/extensions/filters/http/ext_proc/processor_state.cc
Comment (yanjunxiang-google) [inline source/extensions/filters/http/ext_proc/processor_state.cc:388]: same as above

**Comment by yanjunxiang-google** [comment]
@gbrail  Could you please take a look?

**Comment by wbpcode** [comment]
cc @tyxia could you also take a look when you have free time? Thanks.

**Comment by yanavlasov** [comment]
@yanjunxiang-google can you update the doc in the proto file with the new restrictions on using the `mode_override` field, please?

/wait-any

File: source/extensions/filters/http/ext_proc/processor_state.h
Comment (yanjunxiang-google) [inline source/extensions/filters/http/ext_proc/processor_state.h:44]: remove absl::optional based on a comment in a previous PR.

**Comment by yanjunxiang-google** [comment]
> @yanjunxiang-google can you update the doc in the proto file with the new restrictions on using the `mode_override` field, please?
> 
> /wait-any

done

**Comment by yanjunxiang-google** [comment]
@envoyproxy/api-shepherds The API change is only in document.  PTAL

**Review by yanavlasov — DISMISSED** [review]

**Comment by yanjunxiang-google** [comment]
Kind ping

**Review by mattklein123 — APPROVED** [review]

**Review by tyxia — COMMENTED** [review]
I am wondering will this PR touch anything related to STREAM mode? (STREAM mode will also use chunk queue in some cases)

**Comment by yanjunxiang-google** [comment]
> I am wondering will this PR touch anything related to STREAM mode? (STREAM mode will also use chunk queue in some cases)

It doesn't change the normal STREAMED mode behavior.
However, it avoids changing the STREAMED mode to some other modes if data streaming already started.

**Comment by tyxia** [comment]
> > I am wondering will this PR touch anything related to STREAM mode? (STREAM mode will also use chunk queue in some cases)
> 
> 
> 
> It doesn't change the normal STREAMED mode behavior.
> 
> It avoids changing the STREAMED mode to some other modes if data streaming already started.

If it is ok(i.e., this pr is not urgent) I hope to hold this PR a bit as I think there could be a hole with existing chunk queue and stream mode. It could totally unrelated to this PR but I feel we may want to address existing issue first as this pr touches STREAM and chunk queues. I will share more details when I have a chance to look into this.

**Comment by yanjunxiang-google** [comment]
> > > I am wondering will this PR touch anything related to STREAM mode? (STREAM mode will also use chunk queue in some cases)
> > 
> > 
> > It doesn't change the normal STREAMED mode behavior.
> > It avoids changing the STREAMED mode to some other modes if data streaming already started.
> 
> If it is ok(i.e., this pr is not urgent) I hope to hold this PR a bit as I think there could be a hole with existing chunk queue and stream mode. It could totally unrelated to this PR but I feel we may want to address existing issue first as this pr touches STREAM and chunk queues. I will share more details when I have a chance to look into this.


This PR is to clean up the holes in the chunk queue in streamed mode due to processing mode override.
Sure, please share more details when you have chance. I can check whether that hole is fixed or not. If you think this PR bring in new logical issues, please share.

**Review by tyxia — COMMENTED** [review]
Thanks for working on this refactor and clean-up!

With this restriction, removing `delivered_one` related stuff makes things much simpler!

**Comment by tyxia** [comment]
BTW, i have figured out the 

> > > > I am wondering will this PR touch anything related to STREAM mode? (STREAM mode will also use chunk queue in some cases)
> > > 
> > > 
> > > It doesn't change the normal STREAMED mode behavior.
> > > It avoids changing the STREAMED mode to some other modes if data streaming already started.
> > 
> > 
> > If it is ok(i.e., this pr is not urgent) I hope to hold this PR a bit as I think there could be a hole with existing chunk queue and stream mode. It could totally unrelated to this PR but I feel we may want to address existing issue first as this pr touches STREAM and chunk queues. I will share more details when I have a chance to look into this.
> 
> This PR is to clean up the holes in the chunk queue in streamed mode due to processing mode override. Sure, please share more details when you have chance. I can check whether that hole is fixed or not. If you think this PR bring in new logical issues, please share.

Thanks for patience! I have figured out the issues i have seen in HTTP1 environment. It is test issue (i.e., false alarm), which is good :) 
Basically,  there are two passes from connection manager : (1) decodeData("actual_body", end_stream=false) (2) decodeData("empty_body", end_stream = true). Therefore, the test need to processBodyResponse twice (one more time for empty body decodeData to clean up all the entries in the queue. 

It is probably a good idea to enable HTTP1 test environment for ext_proc broadly, after my PR which introduce HTTP1.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:332]: enqueueStreamingChunk() will move the data in tha consolidate buffer: received_data_
https://github.com/envoyproxy/envoy/blob/21af65095760db381347c031971aa5b3668b8968/source/extensions/filters/http/ext_proc/processor_state.h#L57
There is no easy way to read the last piece out without another move() or copyOut() operation.  The proposed approach has least memory operations.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:292]: Yeah, I had been trying to clean up end_stream flag and delivered flag.

delivered flag is completely removed with this PR.

end_stream flag is set the to correct value.  In some cases, like here, the end_stream is true based on the code logic, so, I left it as it is.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:462]: The end_stream flag for body is false if there is trailer after it.

File: source/extensions/filters/http/ext_proc/ext_proc.h
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.h:287]: Thanks for the comments.

The goal of this check is to ignore mode override when receiving none-header response messages.
send_headers_ is a filter configuration, it will always be true in the filter lifetime if header mode is SEND.  
For example, if ext_proc is configured with header SEND and boy STREAMED,  when receiving the STREAMED body response, send_headers_ will be true. That make it not applicable here.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (tyxia) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:332]: `enqueueStreamingChunk` calls ` chunk_queue_.push`, then it calls `queue_.push_back`, where `queue_` is std::deque<QueuedChunkPtr> queue_.  So, I am not sure I fully understand why we cannot read the data from that deque.

Also, here we read the data from the queue right after the `enqueueStreamingChunk` , which  should not involve consolidate. Did I miss something?

Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:332]: The chunk queue only maintain a logical view of the data.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (tyxia) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:462]: I see. so, this end_stream flag represents the same meaning as the end_stream function arg passed to decode/encode callback.  I thought it represent the end of the message sent to ext_proc server.
Thanks!

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (tyxia) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:292]: Then could you elaborate why here `end_stream` is true? We know that there won't be other bodies or trailers after  it?

Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:292]: This is because the if (end_stream) check above.

File: source/extensions/filters/http/ext_proc/ext_proc.h
Comment (tyxia) [reply source/extensions/filters/http/ext_proc/ext_proc.h:287]: I see. Then, should we move whole mode override logic/code to `handleHeaderResponse`?
That is the natural place representing header response handling and we also don't need this check anymore.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (tyxia) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:292]: Then we should just use `end_stream` here? I think we should avoid hardcode if possible. 
I feel it is not good for readability to let people look at code 10 lines above and then figure out value of `end_stream` and its relationship here.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:332]: could you please elaborate where is the one more copy?

File: source/extensions/filters/http/ext_proc/ext_proc.h
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.h:287]: There are two issues to move inside handleHeadersResponse:

1) If we want to do it that way, need to pass in the ProcessResponse, instead of request_headers/response_headers, since mode_override is outside of the HeadersResponse.
2)  Need to add in logic in handleHeadersResponse() to retrieve either response->request_headers() or response->response_headers().  I would rather keep that logic outside in onReceiveMessage() which it is now.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (tyxia) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:332]: ah my bad. I misread your first comment that `setupBodyChunk(), which is copy the data into message` and thought there is an additional copy.

Just took a closer look at code change: basically that copy is to build the request message and already exists in the code. Here you only split the operations

File: source/extensions/filters/http/ext_proc/ext_proc.h
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.h:287]: It might be doable to add a separate mode_override parameter into handleHeadersResponse().  I do not have strong opinion in this way or the current approach.

Comment (tyxia) [reply source/extensions/filters/http/ext_proc/ext_proc.h:287]: Will below work?
Move `decoding_state_.setProcessingMode(response->mode_override())` before line#639 and move `encode` before line 642.   Or something similar to that like checking `response->response_case()`

Keeping it in this file SGTM, but I feel `inHeaderProcessState` is redundant here because we know what kinds of response we are handling.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (tyxia) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:292]: I hope we can use passed-in `end_stream` whenever feasible and have document whenever the hardcoded value is used.

But I think we can do this in a separate PR, do you mind leaving a TODO for this

**Review by tyxia — APPROVED** [review]
LGTM, modulo two nits: 1) removing `inHeaderProcessState` 2) const bool

Thanks for patience, looks great

File: source/extensions/filters/http/ext_proc/ext_proc.h
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.h:287]: BTW, the "response" gives us the response type, but it doesn't tell us the ext_proc filter state.  So, even handleHeadersResponse still need to check the state as here:
https://github.com/envoyproxy/envoy/blob/186cabc53948a89be53e653c78b6ed749495204c/source/extensions/filters/http/ext_proc/processor_state.cc#L86

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:462]: Yeah, that's right.

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:332]: Yeah, that's right.

**Comment by yanjunxiang-google** [comment]
> Thanks for you contribution. Seems like this change the behavior of exist config. So i think a change log is necessary.
> 
> Friendly ping @mattklein123 for API pass.

Done

File: source/extensions/filters/http/ext_proc/ext_proc.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.cc]: Sure, we can use end_stream here since it is available.

File: source/extensions/filters/http/ext_proc/ext_proc.h
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.h]: done

**Comment by yanjunxiang-google** [comment]
> LGTM, modulo two nits: 1) removing `inHeaderProcessState` 2) const bool
> 
> Thanks for patience, looks great

Fixed the other two, but have to keep the inHeaderProcessState() due to the comments above. 
Thanks for the review!

File: source/extensions/filters/http/ext_proc/ext_proc.h
Comment (tyxia) [reply source/extensions/filters/http/ext_proc/ext_proc.h:287]: I think handleHeadersResponse is for both INBOUND and OUTBOUND direction, rather than particular state. It is same as `onReceiveMessage` here. As long as `handleHeadersResponse` is called, we are fine here.  

From code, check seems for spurious message, though it will be safe to override mode in inside check.

Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.h:287]: Yeah, that's right for the first part. For the spurious message check, it is trying to protect Envoy from misbehaved ext_proc server.

**Comment by yanjunxiang-google** [comment]
Kind Ping! The API change is only in documentation.

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Review by mattklein123 — APPROVED** [review]

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
